container-engine-lib/lib/backend_impls/kubernetes/kubernetes_kurtosis_backend/user_services_functions/persistent_directories.go

```diff
@@ -0,0 +1,152 @@
+	return &apiv1.Volume{
+		Name: volumeAndClaim.VolumeName,
+		VolumeSource: apiv1.VolumeSource{
+			HostPath:             nil,
+			EmptyDir:             nil,
+			GCEPersistentDisk:    nil,
+			AWSElasticBlockStore: nil,
+			GitRepo:              nil,
+			Secret:               nil,
+			NFS:                  nil,
+			ISCSI:                nil,
+			Glusterfs:            nil,
+			PersistentVolumeClaim: &apiv1.PersistentVolumeClaimVolumeSource{
+				ClaimName: volumeAndClaim.VolumeClaimName,
+				ReadOnly:  false,
+			},
+			RBD:                  nil,
+			FlexVolume:           nil,
+			Cinder:               nil,
+			CephFS:               nil,
+			Flocker:              nil,
+			DownwardAPI:          nil,
+			FC:                   nil,
+			AzureFile:            nil,
+			ConfigMap:            nil,
+			VsphereVolume:        nil,
+			Quobyte:              nil,
+			AzureDisk:            nil,
+			PhotonPersistentDisk: nil,
+			Projected:            nil,
+			PortworxVolume:       nil,
+			ScaleIO:              nil,
+			StorageOS:            nil,
+			CSI:                  nil,
+			Ephemeral:            nil,
+		},
+	}
+}
+
+func (volumeAndClaim *kubernetesVolumeWithClaim) GetVolumeMount(mountPath string) *apiv1.VolumeMount {
+	return &apiv1.VolumeMount{
+		Name:             volumeAndClaim.VolumeName,
+		ReadOnly:         false,
+		MountPath:        mountPath,
+		SubPath:          "",
+		MountPropagation: nil,
+		SubPathExpr:      "",
+	}
+}
+
+func preparePersistentDirectoriesResources(
+	ctx context.Context,
+	namespace string,
+	serviceUuid service.ServiceUUID,
+	objAttributeProviders object_attributes_provider.KubernetesEnclaveObjectAttributesProvider,
+	serviceMountpointsToPersistentKey map[string]service_directory.DirectoryPersistentKey,
+	kubernetesManager *kubernetes_manager.KubernetesManager,
+) (map[string]*kubernetesVolumeWithClaim, error) {
+	shouldDeleteVolumesAndClaimsCreated := true
+	volumesCreated := map[string]*apiv1.PersistentVolume{}
+	volumeClaimsCreated := map[string]*apiv1.PersistentVolumeClaim{}
+
+	persistentVolumesAndClaims := map[string]*kubernetesVolumeWithClaim{}
+
+	for dirPath, persistentKey := range serviceMountpointsToPersistentKey {
+		volumeAttrs, err := objAttributeProviders.ForSinglePersistentDirectoryVolume(serviceUuid, persistentKey)
+		if err != nil {
+			return nil, stacktrace.Propagate(err, "An error occurred creating the labels for persist service directory '%s'", persistentKey)
+		}
+
+		volumeName := volumeAttrs.GetName().GetString()
+		volumeLabelsStrs := map[string]string{}
+		for key, value := range volumeAttrs.GetLabels() {
+			volumeLabelsStrs[key.GetString()] = value.GetString()
+		}
+
+		var persistentVolume *apiv1.PersistentVolume
+		if persistentVolume, err = kubernetesManager.GetPersistentVolume(ctx, volumeName); err != nil {
+			persistentVolume, err = kubernetesManager.CreatePersistentVolume(ctx, namespace, volumeName, volumeLabelsStrs)
```

Comment:
sorry, but I don't understand the logic here, I think that I'm missing something but, why the persistent volume is created if it already exists? does it work?

Reply:
No we do create it if the `GetPersistentVolume ` returns an error here (`err != nil`), i.e. if it does NOT exist.